search-processors/src/main/java/org/opensearch/searchpipelines/questionanswering/generative/ext/GenerativeQAParameters.java

```diff
@@ -185,7 +187,7 @@ public GenerativeQAParameters(
     public GenerativeQAParameters(StreamInput input) throws IOException {
         this.conversationId = input.readOptionalString();
         this.llmModel = input.readOptionalString();
-        this.llmQuestion = input.readString();
+        this.llmQuestion = input.readOptionalString();
```

Comment:
Hi @austintlee this does not quite make sense to me, why you only make `llmQuestion` into optional while keeping `llmMessage` as mandatory field, if you are trying to make user to choose one of them?

Reply:
@b4sjoo I do make them both optional in the first constructor:

```
public GenerativeQAParameters(
        String conversationId,
        String llmModel,
        String llmQuestion,
        String systemPrompt,
        String userInstructions,
        Integer contextSize,
        Integer interactionSize,
        Integer timeout,
        String llmResponseField,
        List<MessageBlock> llmMessages
    ) {
        this.conversationId = conversationId;
        this.llmModel = llmModel;

        Preconditions
            .checkArgument(
                !(Strings.isNullOrEmpty(llmQuestion) && (llmMessages == null || llmMessages.isEmpty())),
                "At least one of " + LLM_QUESTION + " or " + LLM_MESSAGES_FIELD + " must be provided."
            );
        this.llmQuestion = llmQuestion;
        this.systemPrompt = systemPrompt;
        this.userInstructions = userInstructions;
        this.contextSize = (contextSize == null) ? SIZE_NULL_VALUE : contextSize;
        this.interactionSize = (interactionSize == null) ? SIZE_NULL_VALUE : interactionSize;
        this.timeout = (timeout == null) ? SIZE_NULL_VALUE : timeout;
        this.llmResponseField = llmResponseField;
        if (llmMessages != null) {
            this.llmMessages.addAll(llmMessages);
        }
    }
```

But internally, `llmMessages` is never null and by default is an empty array.

So, when we write out to StreamOut, we don't need to do a null check:
```
public void writeTo(StreamOutput out) throws IOException {
        out.writeOptionalString(conversationId);
        out.writeOptionalString(llmModel);
        out.writeOptionalString(llmQuestion);
        out.writeOptionalString(systemPrompt);
        out.writeOptionalString(userInstructions);
        out.writeInt(contextSize);
        out.writeInt(interactionSize);
        out.writeInt(timeout);
        out.writeOptionalString(llmResponseField);
        out.writeList(llmMessages);
    }
```

Which is why I always expect it to be present (at least as an empty list) when I read it back:
```
public GenerativeQAParameters(StreamInput input) throws IOException {
        this.conversationId = input.readOptionalString();
        this.llmModel = input.readOptionalString();
        this.llmQuestion = input.readOptionalString();
        this.systemPrompt = input.readOptionalString();
        this.userInstructions = input.readOptionalString();
        this.contextSize = input.readInt();
        this.interactionSize = input.readInt();
        this.timeout = input.readInt();
        this.llmResponseField = input.readOptionalString();
        this.llmMessages.addAll(input.readList(MessageBlock::new));
    }
```

Is this an incorrect assumption?  Does the StreamInput constructor need to consider `llmMessages` not being present in `input`?

You can also take a look at stream roundtrip test cases I have in `GenerativeQAParamExtBuilderTests`.

Reply:
It's just because I saw a null check above and then you make here mandatory makes me confused. I think your answer makes sense to me, that `llmMessage` should never be null

Reply:
Besides, changing a `readString()` into `readOptionalString()` could potentially introduce a bwc issue when we have a mixed cluster

Reply:
Going from required to optional should be OK, but not the other way around.  How do we test it?

Reply:
@pyek-bot is currently testing it, he should have a result by tomorrow. Basically we create a lower version cluster (e.g. 2.16) with dedicated master node, then we upgrade the data node to the current version to test. After this we perform the test again, but we upgrade master this time. Does this make sense?

Reply:
If you are changing an optional into mandatory, this is more like a breaking change than breaking bwc, which is more servere :)